lectures/olg.md

```diff
@@ -46,114 +46,895 @@ The OLG model takes up this challenge.
+    α, β, γ, L = model.α, model.β, model.u_params['γ'], model.L
+    lhs = L * (1-α) * (K_prev / L)**α
+    lhs /= (1 + β**(-1/γ) * x**((γ-1)/γ))
+    rhs = L * (x / α)**(1/(α-1))
+    return lhs - rhs
+```
+
+```{code-cell} ipython3
+def find_Rstar(K_prev, model):
+    return optimize.newton(find_Rstar_newton, 0.5, args=(K_prev, model))
+
+def find_Kstar(R_star, model):
+    return model.L * (R_star / model.α)**(1/(model.α-1))
+```
+
+
+The following function plots the equilibrium quantity and equilibrium price.
+
+```{code-cell} ipython3
+def plot_ks_rs(K_t_vals, model):
+    n = len(K_t_vals)
+    R_star = np.empty(n)
+    K_star = np.empty(n)
+
+    for i in range(n):
+        R_star[i] = find_Rstar(K_t_vals[i], model)
+        K_star[i] = find_Kstar(R_star[i], model)
+
+    fig, ax = plt.subplots()
+
+    ax.plot(K_t_vals, R_star, label="equilibrium price")
+    ax.plot(K_t_vals, K_star, label="equilibrium quantity")
+
+    ax.set_xlabel("$K_{t}$")
+    ax.legend()
+    plt.show()
+```
+
+```{code-cell} ipython3
+---
+mystnb:
+  figure:
+    caption: "Equilibrium price and quantity\n"
+    name: equi_ps_q_crra
+  image:
+    alt: equi_ps_q_crra
+    classes: shadow bg-primary
+    width: 200px
+---
+K_t_vals = np.linspace(0.1, 50, 50)
+m_crra = create_olg_model(u=crra, u_params={'γ': 0.5})
+plot_ks_rs(K_t_vals, m_crra)
+```
+
+
+```{solution-end}
+```
+
+```{exercise}
+:label: olg_ex2
+
+Let's keep the model the same except for replacing the utility function $u(c)$ in {eq}`eq_crra`  with a quasilinear form $u(c)=c + c^{\theta}$.
+
+Like what we did in the CRRA case we don't have an analytical solution.
+
+Try to compute the time path capital $\{k_t\}$ in this case.
+```
+
+```{solution-start} olg_ex2
+:class: dropdown
+```
+
+To get the time path capital $\{k_t\}$ first we need to solve the household's utility maximization problem for the optimal consumption and optimal saving.
+
+With the quasilinear preference the Euler equation becomes
+
+```{math}
+:label: euler_quasilinear
+    1 + \theta (w_t - s_t)^{\theta-1} = \beta R_{t+1} + \beta R^{\theta}_{t+1} \theta s_t^{\theta - 1}
+```
+
+Let $k_t := K_t / L$.
+
+Since [](aggregate_supply_capital_log_olg), [](wage_2) and [](interest_rate_2) the Euler equation becomes
+
+```{math}
+:label: euler_quasilinear1
+    1 + \theta ((1-\alpha)k^{\alpha}_t - k_{t+1})^{\theta-1} = \beta \alpha k^{\alpha - 1}_t + \beta (\alpha k^{\alpha - 1}_t)^{\theta} \theta k_{t+1}^{\theta - 1}
+```
+
+Obviously $k_{t+1}$ cannot be solved by pencil and paper.
+
+To solve for $k_{t+1}$ we need to turn to the newton's method.
+
+Let's start by defining the utility function.
+
+```{code-cell} ipython3
+def u_quasilinear(c, θ=4):
+    return c + c**θ
+```
+
+
+The function `find_k_next` is used to find $k_{t+1}$ by finding
+the root of equation [](euler_quasilinear1) using the helper
+function `solve_for_k_next` for a given value of $k_t$.
+
+```{code-cell} ipython3
+def solve_for_k_next(x, k_t, model):
```

Comment:
I got some issues while solving the quasilinear case:
1. Newton method fails to find the root with values of $\theta=1.5, 2.5, 3$ etc. because I see that convergence may be not a real number.
2. For this case too, we can see that $k^*$ is a complex number in the last code cell of exercise 2.